src/api/spec/controllers/webui/users/tokens_controller_spec.rb

```diff
@@ -0,0 +1,169 @@
+require 'rails_helper'
+
+RSpec.describe Webui::Users::TokensController, type: :controller do
+  let(:user) { create(:confirmed_user, login: 'foo') }
+  let(:other_user) { create(:confirmed_user, login: 'bar') }
+
+  before do
+    Flipper[:trigger_workflow].enable
+  end
+
+  describe 'GET #index' do
+    context 'user is not logged in' do
+      it_behaves_like 'require logged in user' do
+        let(:method) { :get }
+        let(:action) { :index }
+      end
+    end
+
+    context 'user is logged in' do
+      before do
+        create(:service_token, user: user)
+        create(:workflow_token, user: user)
+        create(:rss_token, user: user)
+        create(:release_token, user: other_user)
+
+        login user
+        get :index
+      end
+
+      it { expect(response).to render_template('webui/users/tokens/index') }
+      it { expect(assigns(:tokens).count).to eq(2) }
+    end
+  end
+
+  describe 'GET #new' do
+    context 'user is not logged in' do
+      it_behaves_like 'require logged in user' do
+        let(:method) { :get }
+        let(:action) { :new }
+      end
+    end
+
+    context 'user is logged in' do
+      before do
+        login user
+        get :new
+      end
+
+      it { expect(response).to render_template('webui/users/tokens/new') }
+    end
+  end
+
+  describe 'POST #create' do
+    let(:form_parameters) { { token: { operation: 'runservice' } } }
+
+    context 'user is not logged in' do
+      it_behaves_like 'require logged in user' do
+        let(:method) { :post }
+        let(:action) { :create }
+        let(:opts) { { params: form_parameters } }
+      end
+    end
+
+    context 'user is logged in' do
+      RSpec.shared_examples 'check for flashing an error' do
+        it "doesn't flash a success" do
+          expect(subject.request.flash[:success]).to be_nil
+        end
+
+        it 'flashes an error' do
+          expect(subject.request.flash[:error]).not_to be_nil
+        end
+      end
+
+      RSpec.shared_examples 'check for flashing a success' do
+        it 'flashes a success' do
+          expect(subject.request.flash[:success]).not_to be_nil
+        end
+
+        it "doesn't flash an error" do
+          expect(subject.request.flash[:error]).to be_nil
+        end
+      end
+
+      let(:project) { create(:project) }
+      let(:package) { create(:package, project: project) }
+
+      before do
+        login user
+      end
+
+      subject { post :create, xhr: true, params: form_parameters }
+
+      context 'no project and package parameters' do
+        let(:form_parameters) { { token: { operation: 'release' } } }
+
+        include_examples 'check for flashing a success'
```

Comment:
I would feel more comfortable if you would test that the things you want to change, change. Instead of testing this only from the "outside" via the flash.

Reply:
I added tests for checking the number of tokens. Please have a look again.